File: docs/algorithms/code/algorithms/sortedSquares2.ts
function sortedSquares(nums: number[]): number[] {
  let left = 0, right = nums.length - 1;
  const res: number[] = []
  let index = nums.length - 1;
  while(left < right) {
    const leftVal = nums[left] * nums[left];
    const rightVal = nums[right] * nums[right];

    if(leftVal >= rightVal) {
      res.unshift(leftVal);
      left++;
    } else {
      res.unshift(rightVal);
      right--;
    }
  }
  
res.unshift(nums[left] * nums[left])
  return res;
};

File: docs/algorithms/code/algorithms/sortedSquares2.go
// 思路：从两边向中间遍历，每次取两边中较大的数，然后将其放入结果数组中
// 复杂度分析：时间复杂度 O(n), 空间复杂度 O(n)
func sortedSquares(nums []int) []int {
	n := len(nums)
	res := make([]int, n)

	left, right := 0, n-1
	index := n - 1

	for left <= right {
		currLeft := nums[left] * nums[left]
		currRight := nums[right] * nums[right]

		if currLeft > currRight {
			res[index] = currLeft
			left++
		} else {
			res[index] = currRight
			right--
		}
		index--
	}

	return res
}